credsweeper/rules/rule.py

```diff
@@ -37,72 +42,91 @@ class Rule:
+        self.__patterns = Rule._get_patterns(self.rule_type, rule_dict[Rule.VALUES])
+        # auxiliary fields
+        self.__filters = self._get_filters(rule_dict.get(Rule.FILTER_TYPE))
+        self.__pattern_type = Rule._get_pattern_type(self.rule_type, len(self.patterns))
+        self.__use_ml = bool(rule_dict.get(Rule.USE_ML))
+        self.__validations = self._get_validations(rule_dict.get(Rule.VALIDATIONS))
+        self.__required_substrings = [i.strip().lower() for i in rule_dict.get(Rule.REQUIRED_SUBSTRINGS, [])]
+        self.__min_line_len = int(rule_dict.get(Rule.MIN_LINE_LEN, MAX_LINE_LENGTH))
+        self.__usage_list: List[str] = rule_dict.get(Rule.USAGE_LIST, [])
+
+    def _malformed_rule_error(self, rule_dict: Dict, field: str):
+        raise ValueError(f"Malformed rule '{self.__rule_name}'."
+                         f" field '{field}' has invalid value"
+                         f" '{rule_dict.get(field)}'")
+
+    @cached_property
     def rule_name(self) -> str:
         """rule_name getter"""
         return self.__rule_name
 
-    @rule_name.setter
-    def rule_name(self, rule_name: str) -> None:
-        """rule_name setter"""
-        self.__rule_name = rule_name
-
-    @property
+    @cached_property
     def rule_type(self) -> RuleType:
         """rule_type getter"""
         return self.__rule_type
 
-    @rule_type.setter
-    def rule_type(self, rule_type: RuleType) -> None:
-        """rule_type getter"""
-        self.__rule_type = rule_type
-
-    @property
+    @cached_property
     def severity(self) -> Severity:
         """severity getter"""
         return self.__severity
 
-    @severity.setter
-    def severity(self, severity: str) -> None:
-        """severity setter"""
-        if severity_obj := Severity.get(severity):
-            self.__severity = severity_obj
-        else:
-            raise ValueError(f'Malformed rule config file. Rule severity "{severity}" is invalid.')
-
-    @property
+    @cached_property
     def filters(self) -> List[Filter]:
         """filters getter"""
         return self.__filters
 
-    @filters.setter
-    def filters(self, filter_type: str) -> None:
+    def _get_filters(self, filter_type: Union[None, str, List]) -> List[Filter]:
         """filters setter"""
         if filter_type == "" or filter_type is None:
-            self.__filters = []
-        else:
+            return []
+        elif isinstance(filter_type, str):
+            # group applying
             filter_group = getattr(group, filter_type, None)
-            if filter_group is None:
-                raise ValueError(f'Malformed rule config file. Rule filter_type "{filter_type}" is invalid.')
-            self.__filters = filter_group(self.config).filters
+            if isinstance(filter_group, type) and issubclass(filter_group, Group):
+                return filter_group(self.config).filters  # type: ignore
+        elif isinstance(filter_type, list):
```

Comment:
IMO `filter_type` can be changed to `Optional[None]` parameter.
For now it needs duplicated if-else chain.
What do you think?